src/sql/arrow_sql_gen/mysql.rs

```diff
@@ -235,34 +232,66 @@ pub fn rows_to_arrow(rows: &[Row], projected_schema: &Option<SchemaRef>) -> Resu
-                        continue;
-                    };
-
-                    let Some(val) = to_decimal_128(&val, scale) else {
-                        return FailedToConvertBigDecimalToI128Snafu { big_decimal: val }.fail();
-                    };
-
-                    builder.append_value(val);
+                    match arrow_field.data_type() {
+                        DataType::Decimal128(_, _) => {
+                            let Some(builder) = builder.as_any_mut().downcast_mut::<Decimal128Builder>()
+                            else {
+                                return FailedToDowncastBuilderSnafu {
+                                    mysql_type: format!("{mysql_type:?}"),
+                               }
+                               .fail();
+                           };
+                           let val = handle_null_error(row.get_opt::<BigDecimal, usize>(i).transpose())
+                               .context(FailedToGetRowValueSnafu {
+                                    mysql_type: ColumnType::MYSQL_TYPE_DECIMAL,
+                                })?;
+       
+                            let scale = match &val {
+                                Some(val) => val.fractional_digit_count(),
+                                None => 0,
+                            };
+       
+                            let Some(val) = val else {
+                                builder.append_null();
+                                continue;
+                            };
+       
+                            let Some(val) = to_decimal_128(&val, scale) else {
+                                return FailedToConvertBigDecimalToI128Snafu { big_decimal: val }.fail();
+                            };
+       
+                            builder.append_value(val);
+                        }
+                        DataType::Decimal256(_, _) => {
+                            let Some(builder) = builder.as_any_mut().downcast_mut::<Decimal256Builder>()
+                            else {
+                                return FailedToDowncastBuilderSnafu {
+                                    mysql_type: format!("{mysql_type:?}"),
+                                }
+                                .fail();
+                            };
+        
+                            let val = handle_null_error(row.get_opt::<BigDecimal, usize>(i).transpose())
+                                .context(FailedToGetRowValueSnafu {
+                                    mysql_type: ColumnType::MYSQL_TYPE_DECIMAL,
+                                })?;
+        
+                            let Some(val) = val else {
+                                builder.append_null();
+                                continue;
+                            };
+        
+                            let val = to_decimal_256(&val);
+        
+                            builder.append_value(val);
+                        }
+                        _ => unreachable!()
```

Comment:
Why is this unreachable?

Reply:
In this case, the datatype could only be either Decimal128 or Decimal256 given the mapping of `ColumnType::MYSQL_TYPE_DECIMAL` and `ColumnType::MYSQL_TYPE_NEWDECIMAL` in `map_column_to_data_type`. Would it be better to use a error here even if it's not logically unreachable?

Reply:
No, I think its fine to panic for things that can't happen, since that happening would be bug that we hopefully catch in testing. But it would be nice to add a message explaining why its unreachable so that we know what our assumptions were if this was to happen.

Reply:
Got it, I added a comment explaining why it is unreachable